presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -492,6 +493,99 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+                    rightPredicate = leftOuterJoinPushDownResult.getInnerJoinPredicate();
+                    postJoinPredicate = leftOuterJoinPushDownResult.getPostJoinPredicate();
+                    newJoinPredicate = leftOuterJoinPushDownResult.getJoinPredicate();
+                    break;
+                default:
+                    throw new UnsupportedOperationException("Unsupported spatial join type: " + node.getType());
+            }
+
+            newJoinPredicate = simplifyExpression(newJoinPredicate);
+            // TODO: find a better way to directly optimize FALSE LITERAL in join predicate
+            if (newJoinPredicate.equals(BooleanLiteral.FALSE_LITERAL)) {
+                newJoinPredicate = new ComparisonExpression(ComparisonExpression.Operator.EQUAL, new LongLiteral("0"), new LongLiteral("1"));
+            }
+
+            PlanNode leftSource = context.rewrite(node.getLeft(), leftPredicate);
+            PlanNode rightSource = context.rewrite(node.getRight(), rightPredicate);
+
+            PlanNode output = node;
+            if (leftSource != node.getLeft() ||
+                    rightSource != node.getRight() ||
+                    !expressionEquivalence.areExpressionsEquivalent(session, newJoinPredicate, joinPredicate, types)) {
+                // Create identity projections for all existing symbols
+                Assignments.Builder leftProjections = Assignments.builder();
+                leftProjections.putAll(node.getLeft()
+                        .getOutputSymbols().stream()
+                        .collect(Collectors.toMap(key -> key, Symbol::toSymbolReference)));
+
+                Assignments.Builder rightProjections = Assignments.builder();
+                rightProjections.putAll(node.getRight()
+                        .getOutputSymbols().stream()
+                        .collect(Collectors.toMap(key -> key, Symbol::toSymbolReference)));
+
+                leftSource = new ProjectNode(idAllocator.getNextId(), leftSource, leftProjections.build());
+                rightSource = new ProjectNode(idAllocator.getNextId(), rightSource, rightProjections.build());
+
+                output = new SpatialJoinNode(
+                        node.getId(),
+                        node.getType(),
+                        leftSource,
+                        rightSource,
+                        ImmutableList.<Symbol>builder()
+                                .addAll(leftSource.getOutputSymbols())
+                                .addAll(rightSource.getOutputSymbols())
+                                .build(),
+                        newJoinPredicate);
+            }
+
+            if (!postJoinPredicate.equals(TRUE_LITERAL)) {
+                output = new FilterNode(idAllocator.getNextId(), output, postJoinPredicate);
+            }
+
+            output = new ProjectNode(idAllocator.getNextId(), output, Assignments.identity(node.getOutputSymbols()));
```

Comment:
why is this extra identity projection needed?

Reply:
@kokosing That's a good question. This logic was added in https://github.com/mbasmanova/presto/commit/84e9bab3302d043a3fe308668ff9016bba5daecf to fix #8673 and #8682 . I don't really understand why it is needed. @martint Martin, do you know? Would you explain? The tests added along the fix - `TestLogicalPlanner#testLeftConvertedToInnerInequalityJoinNoEquiJoinConjuncts` and `AbstractTestQueries#testNonEqualityLeftJoin` - seem to pass with and without this logic.

Reply:
If I remember correctly, this was needed to preserve the outputs of the join node when a join with non-equi criteria was turned into a join with a filter on top and some underlying columns were only used by the filter. Probably something like:

```
- Join (on f(a)) :: [b, c]
  - t1
  - t2
```

=>

```
- Project :: [b, c]
  - Filter (f(a)) :: [a, b, c]
    - Join :: [a, b, c]
      - t1
      - t2
```

Reply:
`JoinNode` contains a list of output symbols to preserve this information.

Reply:
> but if the filter is pulled out 

Then the proper projection (with only join output symbols) should be added

Reply:
@kokosing That's what I think this code is doing. Identity projections are created for output symbols of the *original* join node: `node.getOutputSymbols()`. However, none of this logic is needed for spatial joins. Hence, I took it out.